### PR TITLE
[android] run a subset of lldb tests against ds2 on Android x86_64 emulator

### PR DESCRIPTION
**Pull request by andrurogerz**
This set of patches adds a new test workflow to the ds2 project. This workflow currently contains a single job which executes a sub-set of lldb tests against ds2 running on an x86_64 Android emulator.

### Notes
1. **The job only runs a very small subset of the lldb API test suite as a smoke test** (test cases found under llvm-project/lldb/test/API/api). We will expand this set of tests after running locally and comparing with test results running against lldb-server instead of ds2. A lot of the existing tests fail even when run against lldb-server, so we will need to add allow/deny lists to filter what test cases can currently run successfully.
2. **The lldb tests require an outdated NDK.** All NDKs after version r21e have a different structure that don't work with the lldb test Android makefile rules. We will need to fix and upstream changes to llvm to get the lldb tests working with later NDKs. In the meantime, using the old NDK is probably sufficient.
3. **The job takes a few hours to run because it has to build clang+lldb from source.** This is unfortunate, but there's really no way to avoid it if we want to run the most up-to-date lldb tests. *Now includes sccache which may improve the llvm build time.*
4. **The job requires more disk space than provided.** A job step to remove some unneeded packages and remove the runner tool cache frees up 10GiB which is sufficient.

### Test Plan
Verified by manually triggering the workflow [here](https://github.com/andrurogerz/ds2/actions/runs/10064785052).

### CODE REVIEW

**Comment by andrurogerz** [comment]
cc @compnerd

The new "test" job will take a few more hours to complete. It *should* be green based on previous runs, but I've made a few tweaks since then so we will see what actually happens.

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        # available GitHub test runners.
+        abi: ['x86_64']
+
+    env:
+      ANDROID_API_LEVEL: 28
+      DEBUG_PORT: 5432
+
+    steps:
+      - uses: actions/checkout@v4
+      - uses: seanmiddleditch/gha-setup-ninja@master
```
Comment (compnerd) [inline .github/workflows/test.yml:27]: Can we pin to a revision instead? The branch name pinning is considered a supply chain attack vector.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+      # Free up some disk space by removing some things we don't use. Without
+      # this step, we exhaust disk space while installing Android the emulator
+      # image in the setup-ndk step below.
+      - name: free-up disk space
+        run: |
+          df -h /home                                                 && \
+          sudo apt purge -q dotnet-* netstandard-* ghc ghc-* boost-*  && \
+          sudo rm -rf "$AGENT_TOOLSDIRECTORY"                         && \
```
Comment (compnerd) [inline .github/workflows/test.yml:36]: This seems ... sketchy. This is the tools cache, are we not using any tools from here?

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+      - name: Build ds2
+        run: |
+          cmake --build ${{ github.workspace }}/BinaryCache/ds2               \
+                --config Release
+
+      # Build clang and a Python-enabled lldb binary to execute the lldb tests.
+      - uses: actions/checkout@v4
+        name: checkout llvm project
+        with:
+          repository: llvm/llvm-project.git
```
Comment (compnerd) [inline .github/workflows/test.yml:58]:
```suggestion
          repository: llvm/llvm-project
```

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+          ref: main
+          path: ${{ github.workspace }}/llvm-project
+      - name: Configure lldb build
+        run: |
+          cmake -B ${{ github.workspace }}/llvm-project/build                   \
+            -S ${{ github.workspace }}/llvm-project/llvm                        \
+            -D LLVM_ENABLE_PROJECTS='clang;lldb'                                \
+            -D LLDB_ENABLE_PYTHON=On                                            \
+            -D CMAKE_BUILD_TYPE=Release                                         \
+            -G Ninja
```
Comment (compnerd) [inline .github/workflows/test.yml:68]:
```suggestion
              -S ${{ github.workspace }}/llvm-project/llvm                      \
              -D LLVM_ENABLE_PROJECTS='clang;lldb'                              \
              -D LLDB_ENABLE_PYTHON=On                                          \
              -D CMAKE_BUILD_TYPE=Release                                       \
              -G Ninja
```

File: .github/workflows/test.yml (hunk 2 above)
Comment (andrurogerz) [reply .github/workflows/test.yml:36]: I don't love it, but the jobs do all run even after deleting this. It is something like 6GiB. TBH, I haven't tried without it, though, so let me do that and see if just uninstalling the packages frees up sufficient space.

Comment (andrurogerz) [reply .github/workflows/test.yml]: I removed this line; will see if the new job run has enough space to pass

Comment (andrurogerz) [reply .github/workflows/test.yml]: Yeah, tests failed; not enough disk space to run the emulator without removing the `$AGENT_TOOLSDIRECTORY` path. Let me see if there are some other packages I can uninstall instead to reclaim a bit more space.

Comment (andrurogerz) [reply .github/workflows/test.yml]: Latest revision uninstalls a bunch more packages (such as web browsers...) and frees up about 7GiB. Hopefully that's enough for the tests to proceed this time.

**Comment by andrurogerz** [comment]
Updated to include sccache ([169fe33](https://github.com/compnerd/ds2/pull/92/commits/169fe332b237bba57b1187373cec1c82f2e62d47)). Hopefully that speeds up build times.

**Comment by compnerd** [comment]
>   ERROR   | Not enough space to create userdata partition. Available: 5069.660156 MB at /home/runner/.android/avd/../avd/test.avd, need 7372.800000 MB.

Seems that we are still a few gigabytes short :(

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+                            microsoft-edge-stable \
+                            google-chrome-stable \
+                            firefox \
+                            dotnet-* \
+                            mono-* \
+                            mysql-* \
+                            postgresql-* \
+                            containerd.io \
+                            docker-ce*
+          sudo apt autoremove
```
Comment (compnerd) [inline .github/workflows/test.yml:52]:
```suggestion
          rm -rf /usr/share/swift
          sudo apt autoremove
```

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        run: |
+          df -h /home
+          sudo apt purge -q azure-cli \
+                            google-cloud-cli* \
+                            powershell \
+                            microsoft-edge-stable \
+                            google-chrome-stable \
+                            firefox \
+                            dotnet-* \
+                            mono-* \
```
Comment (compnerd) [inline .github/workflows/test.yml:47]:
```suggestion
                            mongodb* \
                            mono-* \
```

File: .github/workflows/test.yml [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+                            microsoft-edge-stable \
+                            mono-* \
+                            monodoc-* \
+                            msbuild \
+                            mysql-* \
+                            podman \
+                            postgresql-* \
+                            powershell \
+                            r-base-core \
+                            vim-runtime
```
Comment (compnerd) [inline .github/workflows/test.yml:62]: 😭 (`vim-runtime` should be pretty small, its just a collection of mostly text files).

File: .github/workflows/test.yml [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+          sudo apt autoremove
```
Comment (compnerd) [inline .github/workflows/test.yml:63]:
```suggestion
          sudo rm -rf /usr/share/swift
          sudo apt autoremove
```

THe Swift toolchain is installed by extracting it to this location, it is not a system package.

**Comment by andrurogerz** [comment]
some additional non-package dirs that I can also free up:
```
3.2G	/var/lib/docker
5.0G	/opt/hostedtoolcache/CodeQL
1.2G	/usr/local/share/powershell
1.1G	/usr/local/lib/node_modules
1.9G	/usr/share/swift
```

@compnerd I see you're already trying /usr/share/swift. That's probably enough, but I don't see any reason to not also blow away /usr/local/share/powershell and /var/lib/docker since the related packages are being removed. Sound reasonable?

**Pull request merged**